src/pkg/artifact/manager.go

```
@@ -30,44 +32,117 @@ type Manager interface {
+	UpdatePullTime(ctx context.Context, artifactID int64, time time.Time) (err error)
 }
 
 // NewManager returns an instance of the default manager
 func NewManager() Manager {
-	return &manager{}
+	return &manager{
+		dao.New(),
+	}
 }
 
 var _ Manager = &manager{}
 
 type manager struct {
+	dao dao.DAO
 }
 
-func (m *manager) List(ctx context.Context, query *q.Query) (total int64, artifacts []*Artifact, err error) {
-	// TODO implement
-	return 0, nil, nil
+func (m *manager) List(ctx context.Context, query *q.Query) (int64, []*Artifact, error) {
+	total, err := m.dao.Count(ctx, query)
+	if err != nil {
+		return 0, nil, err
+	}
+	arts, err := m.dao.List(ctx, query)
+	if err != nil {
+		return 0, nil, err
+	}
+	var artifacts []*Artifact
+	for _, art := range arts {
+		artifact, err := m.assemble(ctx, art)
+		if err != nil {
+			return 0, nil, err
+		}
+		artifacts = append(artifacts, artifact)
+	}
+	return total, artifacts, nil
 }
+
 func (m *manager) Get(ctx context.Context, id int64) (*Artifact, error) {
-	// TODO implement
-	return nil, nil
+	art, err := m.dao.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return m.assemble(ctx, art)
 }
-func (m *manager) Create(ctx context.Context, artifact *Artifact) (id int64, err error) {
-	// TODO implement
-	return 0, nil
+func (m *manager) Create(ctx context.Context, artifact *Artifact) (int64, error) {
+	id, err := m.dao.Create(ctx, artifact.To())
+	if err != nil {
+		return 0, err
+	}
+	for _, reference := range artifact.References {
+		reference.ParentID = id
+		if _, err = m.dao.CreateReference(ctx, reference.To()); err != nil {
+			return 0, err
+		}
+	}
+	return id, nil
 }
 func (m *manager) Delete(ctx context.Context, id int64) error {
-	// TODO implement
-	return nil
+	// check whether the artifact is referenced by other artifacts
+	references, err := m.dao.ListReferences(ctx, &q.Query{
+		Keywords: map[string]interface{}{
+			"child_id": id,
+		},
+	})
+	if err != nil {
+		return err
```

This is where you should return a meaningful error...